core/src/main/scala/kafka/coordinator/TransactionCoordinator.scala

```diff
@@ -0,0 +1,137 @@
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package kafka.coordinator
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.server.{KafkaConfig, ReplicaManager}
+import kafka.utils.{Logging, Pool, ZkUtils}
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.utils.Time
+
+/**
+  * Transaction coordinator handles message transactions sent by producers and communicate with brokers
+  * to update ongoing transaction's status.
+  *
+  * Each Kafka server instantiates a transaction coordinator which is responsible for a set of
+  * producers. Producers with specific transactional ids are assigned to their corresponding coordinators;
+  * Producers with no specific transactional id may talk to a random broker as their coordinators.
+  */
+object TransactionCoordinator {
+
+  def apply(config: KafkaConfig, zkUtils: ZkUtils, time: Time): TransactionCoordinator = {
+
+    val pIDManager = new PidManager(config.brokerId, zkUtils)
+    val logManager = new TransactionLogManager(config.brokerId, zkUtils)
+    new TransactionCoordinator(config.brokerId, pIDManager, logManager)
+  }
+}
+
+class TransactionCoordinator(val brokerId: Int,
+                             val pIDManager: PidManager,
+                             val logManager: TransactionLogManager) extends Logging {
+
+  this.logIdent = "[Transaction Coordinator " + brokerId + "]: "
+
+  type InitPIDCallback = InitPIDResult => Unit
+
+  /* Active flag of the coordinator */
+  private val isActive = new AtomicBoolean(false)
+
+  /* TransactionalId to pid metadata map cache */
+  private val pIDMetadataCache = new Pool[String, PidMetadata]
+
+  def handleInitPID(transactionalId: String,
+                    responseCallback: InitPIDCallback): Unit = {
+    if (transactionalId == null || transactionalId.isEmpty) {
+      // if the transactional id is not specified, then always blindly accept the request
+      // and return a new PID from the PID manager
+      val pID: Long = pIDManager.getNewPid()
+
+      responseCallback(InitPIDResult(pID, epoch = 0, Errors.NONE))
+    } else if(!logManager.isCoordinatorFor(transactionalId)) {
+      // check if it is the assigned coordinator for the transactional id
+      responseCallback(initPIDError(Errors.NOT_COORDINATOR_FOR_GROUP))
+    } else {
+      // only try to get a new PID and update the cache if the transactional id is unknown
+      getPIDMetadata(transactionalId) match {
+        case None =>
+          val pID: Long = pIDManager.getNewPid()
+          val metadata = addPIDMetadata(transactionalId, new PidMetadata(pID))
+
+          responseCallback(initPIDMetadata(metadata))
+
+        case Some(metadata) =>
+          metadata.epoch = (metadata.epoch + 1).toShort
```

Comment:
Do we need to protect this? I'm wondering what our approach to locking will be in general. For consumers, we had the group, but it would be unfortunate if we needed a separate lock for every transactionalId. I could imagine having a fixed set of locks and hashing the transactionalId onto one of them. There are probably better ideas.